doc/analyzers/VSTHRD003.md

```diff
@@ -90,3 +92,116 @@ class SomeClass
+Now suppose that another thread comes along and wants to do some work before calling `UnlockProgress()`:
+
+```cs
+async Task KeyMasterAsync()
+{
+    await joinableTaskFactory.SwitchToMainThreadAsync();
+    // do some work
+    // Unblock others
+    someClass.UnlockProgress();
+}
+```
+
+We have a deadlock, because `SomeCaller()` is blocking the main thread while waiting for `UnlockProgress()` to be called, but `UnlockProgress()` will not be called until `KeyMasterAsync` can reach the main thread.
+
+Fixing this fundamentally means that `SomeCaller` will need to *join* whatever work may be needed to ultimately call `UnlockProgress`. But for `SomeCaller`, that work is unknown, since it's at least partially inside another class.
+`TaskCompletionSource<T>` is fundamentally a blackbox and the most difficult thing to use correctly while avoiding deadlocks.
+If you can instead use `JoinableTaskFactory.RunAsync` to execute your code as given in the solution above, that is ideal.
+
+Assuming you must keep using `TaskCompletionSource<T>` though, here's how it can be done as safely as possible.
+Joining a set of unknown work is best done with the `JoinableTaskCollection` class.
+It is the responsibility of `SomeClass` in the example above to work with this collection to avoid deadlocks, like this:
+
+```cs
+class SomeClass
+{
+    TaskCompletionSource<bool> tcs = new();
+    JoinableTaskCollection jtc = new();
+    JoinableTaskFactory jtf;
+
+    SomeClass()
+    {
+        this.jtf = joinableTaskContext.CreateFactory(this.jtc);
+    }
+
+    public async Task MyMethodAsync()
+    {
+        // Our caller is interested in completion of the TaskCompletionSource,
+        // so join the collected effort while waiting, to avoid deadlocks.
+        using (this.jtc.Join())
+        {
+            await this.tcs.Task; /* This analyzer will report warning on this line. */
+        }
+
+        /* do more stuff */
+    }
+
+    void UnlockProgress()
+    {
+        this.tcs.TrySetResult(true);
+    }
+
+    async Task KeyMasterAsync()
+    {
+        // As this method must complete to signal the TaskCompletionSource,
+        // all of its work must be done within the context of a JoinableTask
+        // that belongs to the JoinableTaskCollection.
+        // jtf.RunAsync will add the JoinableTask it creates to the jtc collection
+        // because jtf was created with jtc as an argument in our constructor.
+        await jtf.RunAsync(async delegate
+        {
+            // Because we're in the jtc collection, anyone waiting on MyMethodAsync
+            // will automatically lend us use of the main thread if they have it
+            // to avoid deadlocks.
+            // It does NOT matter whether we use jtf or another JoinableTaskFactory instance
+            // at this point.
+            await joinableTaskFactory.SwitchToMainThreadAsync();
+
+            // do some work
+            // Unblock others
+            someClass.UnlockProgress();
```

Comment:
at this line, `someClass` is just `this`, since the method is part of `SomeClass`. In the sample above I thought `KeyMasterAsync` method was assumed to be part of another class (because again it used `someClass` instead of `this`.

Reply:
You're right. That wasn't consistent. I've made some corrections.